packages/util-endpoints/src/EndpointResolverCache.ts

```diff
@@ -0,0 +1,90 @@
+import type { EndpointParams, EndpointV2, RuleSetObject } from "@smithy/types";
+
+/**
+ * @internal
+ *
+ * Used by the resolveEndpoint function.
+ */
+export class EndpointResolverCache {
+  private static rulesetMap = new Map<RuleSetObject, EndpointResolverCache>();
+  private store = new Map<string, EndpointV2>();
+
+  public constructor(
+    /**
+     * Desired maximum number of entries in the cache.
+     */
+    public capacity: number = 100
+  ) {}
+
+  /**
+   * @returns the instance associated with a particular ruleset object.
+   */
+  public static forRuleset(ruleset: RuleSetObject): EndpointResolverCache {
+    if (!EndpointResolverCache.rulesetMap.get(ruleset)) {
+      EndpointResolverCache.rulesetMap.set(ruleset, new EndpointResolverCache());
+    }
+    return EndpointResolverCache.rulesetMap.get(ruleset)!;
+  }
+
+  /**
+   * Clear all registered instances and their caches.
+   */
+  public static clearAll(): void {
+    for (const [, cache] of EndpointResolverCache.rulesetMap) {
+      cache.clear();
+    }
+    EndpointResolverCache.rulesetMap.clear();
+  }
+
+  /**
+   * Create a cache key.
+   */
+  public key(endpointParams: EndpointParams): string {
+    let buffer = "";
+    for (const key in endpointParams) {
+      buffer += key;
+      buffer += `=(${typeof endpointParams[key]})`;
+      buffer += endpointParams[key];
+      buffer += ",";
+    }
+    return buffer;
+  }
+
+  /**
+   * Read from the cache.
+   */
+  public get(key: string): EndpointV2 | undefined {
+    return this.store.get(key);
+  }
+
+  /**
+   * Write to the cache.
+   */
+  public set(key: string, endpoint: EndpointV2): void {
+    this.store.set(key, endpoint);
+    this.trim();
```

Comment:
Use `LRUCache` from `mnemonist/lru-cache`?

We currently use it in `endpoint-cache` for flexible endpoints https://github.com/aws/aws-sdk-js-v3/blob/5afac5e175e7b0a92cd612c5a4730a4269c4fa52/packages/endpoint-cache/src/EndpointCache.ts#L1C23-L1C42